cpu/cc3200/periph/spi.c

```diff
@@ -0,0 +1,250 @@
+{
+    /* read configured word length */
+    uint32_t spi_wl = spi(bus)->ch0_conf & MCSPI_CH0CONF_WL_M;
+
+    /* convert spi_wl word length constant to word length in byte: (x >> 7) + 1
+     * returns word length in bits, then (bits >> 3) divides it by 8 and results
+     * in word length in bytes. Combined ((x >> 7) + 1) >> 3 == (x >> 10) + 1
+     */
+    return !(len % ((spi_wl >> 10) + 1));
+}
+
+/**
+ * @brief reset spi to default state
+ *
+ * @param bus spi bus id
+ */
+void spi_reset(spi_t bus)
+{
+    volatile cc3200_spi_t *dev = spi(bus);
+
+    /* disable chip select in software controlled mode */
+    dev->ch0_conf &= ~MCSPI_CH0CONF_FORCE;
+
+    /* Disable SPI Channel */
+    dev->ch0_ctrl &= ~MCSPI_CH0CTRL_EN;
+
+    /* reset SPI */
+    dev->sys_conf |= MCSPI_SYSCONFIG_SOFTRESET;
+
+    /* wait for reset */
+    while (!((dev->sys_status) & MCSPI_SYSSTATUS_RESETDONE)) {
+    }
+
+    /* enable spi */
+    dev->ch0_ctrl &= ~MCSPI_CH0CTRL_EN;
+}
+
+/**
+ * @brief configure spi module. This functions is closely modeled after
+ * SPIConfigSetExpClk.
+ *
+ * @param bus
+ * @param mode SPI operation sub mode
+ * @param clk SPI bit rate
+ */
+void _spi_config(spi_t bus, spi_mode_t mode, spi_clk_t clk)
+{
+    volatile cc3200_spi_t *dev = spi(bus);
+    /* current value of the ctrl register is used as a starting point */
+    cc3200_reg_t tmp = dev->module_ctrl;
+
+    /* compute divider value */
+    cc3200_reg_t divider = (XTAL_CLK / clk) - 1;
+
+    /* set master mode with hardware chip select */
+    tmp &= ~(MCSPI_MODULCTRL_MS | MCSPI_MODULCTRL_SINGLE);
+
+    /* set SPI config */
+    /* TIs code is also using OR with Master/Slave mode */
+    /* since riot only supports master mode we can skip that */
+    tmp |= (spi_config[bus].config >> 24) & 0xFF;
+
+    /* write config */
+    dev->module_ctrl = tmp;
+
+    /* reset tmp and set IS, DPE0, DPE1 for master mode */
+    tmp = 0x1 << 16;
+
+    /* mask config and set clock divider granularity to 1 cycle */
+    tmp = (tmp & ~(MCSPI_CH0CONF_WL_M | MCSPI_CH0CONF_EPOL | MCSPI_CH0CONF_POL |
+                   MCSPI_CH0CONF_PHA | MCSPI_CH0CONF_TURBO)) |
+          MCSPI_CH0CONF_CLKG;
+
+    /* The least significant four bits of the divider is used fo configure */
+    /* CLKD in MCSPI_CHCONF next eight least significant bits are used to */
+    /* configure the EXTCLK in MCSPI_CHCTRL */
+    tmp |= ((divider & 0x0000000F) << 2);
+    dev->ch0_ctrl = ((divider & 0x00000FF0) << 4);
+
+    /* set protocol, CS polarity, word length and turbo mode */
+    dev->ch0_conf = ((tmp | mode) | (spi_config[bus].config & 0x0008FFFF));
+}
+
+void spi_init(spi_t bus)
+{
+    /* assert(bus >= SPI_NUMOF); */
+    mutex_init(&locks[bus]);
+
+    /* CC3100 module does not require pin config */
+    if (bus != CC3100_SPI) {
+        /* trigger pin initialization */
+        spi_init_pins(bus);
+    }
+
+    /* enable clock */
+    switch (bus) {
+    case EXTERNAL_SPI:
```

Comment:
Would that still be true if someone were to swap the two entries in `spi_config`?

Reply:
that would break it... I am unsure how to handle that properly. The internal SPI is used to communicate with the network processor. Should I just number them and add a comment?

Reply:
Would checking the `gpio_port` or `base_addr` property help?

Reply:
Would something like that work 
```
    (*cc3200_arcm_reg_t)(&ARCM->MCSPI_A1 + bus *
                         0x1C)->clk_gating |= spi_config[bus].config &
                                              PRCM_MODE_CLK_MASK;
```
or is there a simple C trick to avoid multiplying by the bus index

Reply:
`bus * 0x1C`? What are you trying to do?

I noticed you have `LSPI_BASE` and `GSPI_BASE` in `base_addr`. Will `LSPI_BASE` always be controlled through `MCSPI_A1`?

Or does `gpio_port == 0` map to `MCSPI_A1` and `gpio_port == 1` to `MCSPI_A2`?

And what's with `PRCM_SLP_MODE_CLK`?

Reply:
`LSPI_BASE` is controlled by `MCSPI_A1` and `GSPI_BASE` is controlled by `MCSPI_A2` so in that sense the later is the case. My idea was that `&ARCM->MCSPI_A1 + bus * 0x1C` will return `MCSPI_A2` when `bus == 1`. I could put it into a separate method to make it more clear. 
I would move `PRCM_SLP_MODE_CLK` to the board specific SPI config and extract the value via the `PRCM_MODE_CLK_MASK`

Reply:
There is no need to get rid of the switch case.
But the `bus` variable will only tell you the index in the `spi_config` struct.

Couldn't you do

```C
if (spi(bus) == GSPI_BASE) {
    ARCM->MCSPI_A1.clk_gating |= PRCM_RUN_MODE_CLK;
else if (spi(bus) == LSPI_BASE) {
     ARCM->MCSPI_A2.clk_gating |= PRCM_RUN_MODE_CLK | PRCM_SLP_MODE_CLK;
}
```

so the index does not matter?

Reply:
No, the index does not matter. If using the address is okay I would just use your suggestion here.